src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java

```diff
@@ -0,0 +1,230 @@
+      Path pathToSrcsList,
+      ImmutableList<String> extraArguments,
+      ImmutableSortedSet<Path> declaredClassPathEntries,
+      Optional<Path> workingDirectory,
+      ProjectFilesystem projectFilesystem)
+      throws InterruptedException {
+
+    String classpath =
+        Joiner.on(File.pathSeparator).join(transform(declaredClassPathEntries, Object::toString));
+
+    ImmutableList.Builder<String> argsBuilder = ImmutableList.<String>builder()
+        .add(INCLUDE_RUNTIME_FLAG)
+        .add(DESTINATION_FLAG)
+        .add(projectFilesystem.resolve(outputDirectory).toString())
+        .addAll(extraArguments)
+        .addAll(transform(
+            kotlinSourceFilePaths,
+            path -> projectFilesystem.resolve(path).toAbsolutePath().toString())
+        );
+
+    if (!classpath.isEmpty()) {
+      argsBuilder
+          .add(CLASSPATH_FLAG)
+          .add(classpath.isEmpty() ? "''" : classpath);
+    }
+
+    ImmutableList<String> args = argsBuilder.build();
+
+    Set<File> compilerIdPaths = compilerClassPath
+        .stream()
+        .map(Path::toFile)
+        .collect(Collectors.toSet());
+
+    try {
+       Object compilerShim = kotlinShims.computeIfAbsent(
+          compilerIdPaths
+              .stream()
+              .map(File::getAbsolutePath)
+              .collect(Collectors.toSet()),
+          k -> loadCompilerShim(context));
+
+      Method compile = compilerShim
+          .getClass()
+          .getMethod("exec", PrintStream.class, String[].class);
+
+      Class<?> exitCodeClass = compilerShim
+          .getClass()
+          .getClassLoader()
+          .loadClass(EXIT_CODE_CLASS);
+
+      Method getCode = exitCodeClass.getMethod("getCode");
+
+      Object exitCode = compile.invoke(
+          compilerShim,
+          new UncloseablePrintStream(context.getStdErr()),
+          args.toArray(new String[0]));
+
+      return (Integer) getCode.invoke(exitCode);
+
+    } catch (IllegalAccessException | InvocationTargetException |
+        NoSuchMethodException | ClassNotFoundException ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  @Override
+  public ImmutableCollection<BuildRule> getDeps(SourcePathRuleFinder ruleFinder) {
+    return ruleFinder.filterBuildRuleInputs(getInputs());
+  }
+
+  @Override
+  public ImmutableCollection<SourcePath> getInputs() {
+    return ImmutableSet.of();
+  }
+
+  @Override
+  public void appendToRuleKey(RuleKeyObjectSink sink) {
+    sink.setReflectively("kotlinc", "jar-backed")
+        .setReflectively("kotlinc.version", "in-memory")
+        .setReflectively("kotlinc.classpath", compilerClassPath.toString());
+  }
+
+  private Object loadCompilerShim(ExecutionContext context) {
+    try {
+      ClassLoaderCache classLoaderCache = context.getClassLoaderCache();
+      classLoaderCache.addRef();
+
+      ClassLoader classLoader = classLoaderCache.getClassLoaderForClassPath(
+          this.getClass().getClassLoader(),
```

Comment:
Found the culprit!

We're running the kotlin compiler with a classpath that includes all the classes loaded for buck.  That's not what we want. The only classes the compiler should know about / have access to are the ones specified in the `compilerClassPath`.

Changing this to `null` allows for successful compilation without needing to change `mapdb` or anything in `third-party`.